### PR TITLE
Fix for live lock during Stage shutdown

### PR DESCRIPTION
**Pull request by johnou**
Motivation:

hosting.notifyStateChange() wrapped in await may cause executionSerializer to shut down within a task executed by executionSerializer.

Modifications:

Ensure Stage shutdown continues from calling thread.

Result:

Execution executor shuts down cleanly.

### CODE REVIEW

**Comment by johnou** [comment]
```
"ForkJoinPool-1-worker-11" prio=10 tid=0x000000cb nid=0x00cb waiting [0x00000000]
   java.lang.Thread.State: TIMED_WAITING
        at java.lang.Object.wait(Native Method)
        - waiting on  <0x58399d82> (a java.util.concurrent.ForkJoinPool)
        at java.util.concurrent.ForkJoinPool.awaitTermination(ForkJoinPool.java:3141)
        at cloud.orbit.actors.concurrent.WaitFreeMultiExecutionSerializer.shutdown(WaitFreeMultiExecutionSerializer.java:110)
        at cloud.orbit.actors.Stage.async$stop(Stage.java:772)
        at cloud.orbit.actors.Stage$$Lambda$128/1753749782.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at cloud.orbit.concurrent.Task.internalComplete(Task.java:122)
        at cloud.orbit.concurrent.Task.lambda$from$1(Task.java:173)
        at cloud.orbit.concurrent.Task$$Lambda$38/687251464.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
        at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at cloud.orbit.concurrent.Task.internalComplete(Task.java:122)
        at cloud.orbit.concurrent.Task.lambda$from$1(Task.java:173)
        at cloud.orbit.concurrent.Task$$Lambda$38/687251464.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
        at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at cloud.orbit.concurrent.Task.internalComplete(Task.java:122)
        at cloud.orbit.concurrent.Task.lambda$from$1(Task.java:173)
        at cloud.orbit.concurrent.Task$$Lambda$38/687251464.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
        at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at cloud.orbit.concurrent.Task.internalComplete(Task.java:122)
        at cloud.orbit.concurrent.Task.lambda$from$1(Task.java:173)
        at cloud.orbit.concurrent.Task$$Lambda$38/687251464.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
        at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:561)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:739)
        at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by johnou** [comment]
Not a complete fix..

```
2017-02-21 21:09:55,828 DEBUG [Thread-3] (cloud.orbit.actors.Stage) Start stopping pipeline
2017-02-21 21:09:56,143 DEBUG [ForkJoinPool-1-worker-15] (cloud.orbit.actors.Stage) Stopping actors
2017-02-21 21:09:57,043 DEBUG [ForkJoinPool-1-worker-15] (cloud.orbit.actors.Stage) Stopping timers
..
2017-02-21 21:09:57,981 INFO  [ThreadDumper] 
```

File: actors/runtime/src/main/java/cloud/orbit/actors/Stage.java [hunk 1]
```diff
@@ -753,28 +754,31 @@ public void addExtension(final ActorExtension extension)
         // * wait pending tasks execution
         // * stop the network
 
-        logger.debug("start stopping pipeline");
-        await(hosting.notifyStateChange());
+        logger.debug("Start stopping pipeline");
+        CompletableFuture<Void> notified = new CompletableFuture<>();
+        // await must not be used directly on actor call otherwise shutdown may continue in execution thread and cause livelock
+        hosting.notifyStateChange().whenCompleteAsync((r, e) -> notified.complete(null));
```
Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/Stage.java]: Might be cleaner to just run ```executionSerializer.shutdown();``` in some fire and forget thread.

**Comment by johnou** [comment]
@JoeHegarty I think this is good to pull in now.

File: actors/runtime/src/main/java/cloud/orbit/actors/Stage.java [hunk 2]
```diff
@@ -1072,15 +1076,15 @@ public void dispose()
             @Override
             public void run()
             {
-                if (localActor.isDeactivated())
+                if (localActor.isDeactivated() || getState() != NodeCapabilities.NodeState.RUNNING)
```
Comment (DanielSperry) [inline actors/runtime/src/main/java/cloud/orbit/actors/Stage.java]: I'd recommend adding some tests to check if this works as intended.
I'm worried with actor calls being received before the node is up.
And also with race conditions while shutting down.

Comment (johnou) [inline actors/runtime/src/main/java/cloud/orbit/actors/Stage.java]: NodeCapabilities.NodeState.RUNNING is set before the node joins the cluster inside cloud.orbit.actors.Stage#start and at that point should not receive any requests until that future completes.

If the node is shutting down I don't think it makes sense to continue running the timer task if it hasn't passed this checkpoint.

I'm happy to revert these two changes to the TimerTask and open a new pull request as they are not 100% related to the lock on shutdown.

**Comment by johnou** [comment]
@DanielSperry PTAL

**Pull request merged**